### PR TITLE
[feature] Allow users to save expression functions in QGIS project file

### PR DESCRIPTION
**Pull request by gacarrillor**
In the expression builder dialog, Python functions can only be saved in the user profile directory. If a user wants to share expression functions with colleagues, he/she needs to share his/her expressions directory or individual `.py` files, which is inconvenient (without mentioning the installation of those shared files won't be trivial either).

This PR allows users to also store expression functions into the current QGIS project file, making it easier to share functions with other interested parties.

Some details about the implementation:

  + At most, one `[Project Functions]` item will be listed in the left-hand side list widget (`Function Editor` tab).
  + Users will be able to add and remove the aforementioned item.
  + If a `[Project Functions]` item already exists in the list, the corresponding option in the `Add Function File` dialog will be disabled.
  + Project functions will be loaded when opening the QGIS project.
  + Project gets dirty when expression functions are added to the project or when they are edited.
  + Note: Project functions are not unloaded for the ongoing QGIS session when a project is closed. This is the same behavior compared to expression functions saved in the user profile (i.e., when a function file is removed, expression functions are not unloaded for the ongoing QGIS session).

image

image

Labels: [Needs Documentation] [Feature] [Expressions]

Funded by the [QGIS user group Switzerland](https://qgis.ch/).

### CODE REVIEW

File: src/gui/qgsexpressionaddfunctionfiledialog.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "qgis_gui.h"
+#include <QDialog>
+#include "ui_qgsexpressionaddfunctionfiledialogbase.h"
+
+
+/**
+ * \ingroup gui
+ * \brief A dialog to select whether to create a function file or project functions.
+ * \since QGIS 3.40
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.h:27]: Question: should we expose this to SIP? Feels like it is serving such a niche scenario here it's not worth doing so. Less API promise to care about.

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#include <QPushButton>
+#include <QStandardItemModel>
+#include "qgsexpressionaddfunctionfiledialog.h"
+
+QgsExpressionAddFunctionFileDialog::QgsExpressionAddFunctionFileDialog( const bool enableProjectFunctions, QWidget *parent )
+  : QDialog( parent )
+{
+  setupUi( this );
+  cboFileOptions->addItem( QStringLiteral( "Function file" ) );
+  cboFileOptions->addItem( QStringLiteral( "Project functions" ), QStringLiteral( "project" ) );
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.cpp]:
```suggestion
  cboFileOptions->addItem( tr( "Function file" ) );
  cboFileOptions->addItem( tr( "Project functions" ), QStringLiteral( "project" ) );
```

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  connect( cboFileOptions, qOverload<int>( &QComboBox::currentIndexChanged ), this, &QgsExpressionAddFunctionFileDialog::cboFileOptions_currentIndexChanged );
+  connect( txtNewFileName, &QLineEdit::textChanged, this, [ = ]( const QString & ) { updateOkButtonStatus(); } );
+
+  updateOkButtonStatus();
+}
+
+void QgsExpressionAddFunctionFileDialog::cboFileOptions_currentIndexChanged( int )
+{
+  bool projectSelected = cboFileOptions->currentData() == "project";
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.cpp]:
```suggestion
  bool projectSelected = cboFileOptions->currentData() == QStringLiteral( "project" );
```

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  txtNewFileName->setVisible( !projectSelected );
+  updateOkButtonStatus();
+}
+
+void QgsExpressionAddFunctionFileDialog::updateOkButtonStatus()
+{
+  QPushButton *okBtn = buttonBox->button( QDialogButtonBox::StandardButton::Ok );
+  okBtn->setEnabled( true );
+
+  if ( cboFileOptions->currentData() != "project" )
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.cpp]: Same as above.

File: src/gui/qgsexpressionaddfunctionfiledialog.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+     * or not, which is useful to avoid the creation of multiple 'Project Functions' items in the
+     * Function Editor file list.
+     * \a parent is the parent widget.
+     */
+    QgsExpressionAddFunctionFileDialog( const bool enableProjectFunctions, QWidget *parent );
+
+    /**
+     * Returns whether user has selected to create project functions
+     */
+    bool createProjectFunctions() const {return cboFileOptions->currentData() == "project";};
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.h]: Try to run the pre-commit script on the header to fix spacing issues? I would move that to the cpp too. Same for the line below.

File: src/gui/qgsexpressionaddfunctionfiledialog.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Creates a QgsExpressionAddFunctionFileDialog to create function files or to set the current
+     * project as a function container.
+     * \a enableProjectFunctions defines whether the dialog will enable the 'Project function' option
+     * or not, which is useful to avoid the creation of multiple 'Project Functions' items in the
+     * Function Editor file list.
+     * \a parent is the parent widget.
+     */
+    QgsExpressionAddFunctionFileDialog( const bool enableProjectFunctions, QWidget *parent );
```
Comment (nirvn) [inline src/gui/qgsexpressionaddfunctionfiledialog.h]: -const

File: src/gui/qgsexpressionbuilderwidget.cpp [hunk 7]
```diff
@@ -443,59 +466,112 @@ void QgsExpressionBuilderWidget::newFunctionFile( const QString &fileName )
...
   {
-    newFunctionFile( text );
+    if ( dlg.createProjectFunctions() )
+    {
+      QListWidgetItem *item = new QListWidgetItem( QgsApplication::getThemeIcon( QStringLiteral( "mIconQgsProjectFile.svg" ) ), DEFAULT_PROJECT_FUNCTIONS_ITEM_NAME );
+      item->setData( Qt::UserRole, QStringLiteral( "project" ) );
+      cmbFileNames->insertItem( 0, item );
+      cmbFileNames->setCurrentRow( 0 );
+
+      QString templatetxt;
```
Comment (nirvn) [inline src/gui/qgsexpressionbuilderwidget.cpp]: I'd avoid acronyms here, expanded wording == easier to read. I.e. here, templateText > templateTxt :)

**Review by nirvn — COMMENTED** [review]
Loving this function, I actually had to avoid using these exactly because they weren't portable / embedded into shared projects. A few comments here and there.

File: src/gui/qgsexpressionaddfunctionfiledialog.h (hunk 1 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.h:27]: Agreed, removed from SIP.

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp (hunk 2 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.cpp]: Applied, thanks!

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp (hunk 3 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.cpp]: Applied, thanks!

File: src/gui/qgsexpressionaddfunctionfiledialog.cpp (hunk 4 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.cpp]: Applied, thanks!

File: src/gui/qgsexpressionaddfunctionfiledialog.h (hunk 5 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.h]: Applied, and moved to the cpp, thanks!

File: src/gui/qgsexpressionaddfunctionfiledialog.h (hunk 6 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionaddfunctionfiledialog.h]: Removed, thanks!

File: src/gui/qgsexpressionbuilderwidget.cpp (hunk 7 above)
Comment (gacarrillor) [reply src/gui/qgsexpressionbuilderwidget.cpp]: Done!

**Comment by gacarrillor** [comment]
> Loving this function, I actually had to avoid using these exactly because they weren't portable / embedded into shared projects. A few comments here and there.

Thanks a lot for your review, all changes were applied in https://github.com/gacarrillor/QGIS/pull/43/commits/e9f736e813599f060cddc0512b3f38ec40318935
I'll create a PR to QGIS repo when tests are green.

**Pull request merged**